aptos-move/move-examples/shared_account/sources/SharedAccount.move

```diff
@@ -0,0 +1,195 @@
+        assert!(Vector::length(&addresses) == Vector::length(&numerators), Errors::invalid_argument(EINVALID_INPUT));
+        assert!(Vector::length(&addresses) == Vector::length(&denominators), Errors::invalid_argument(EINVALID_INPUT));
+
+        assert!(Account::exists_at(Signer::address_of(source)), Errors::invalid_argument(EACCOUNT_NOT_FOUND));
+        let (resource_signer, resource_signer_cap) = Account::create_resource_account(source, seed);
+        if (!exists<SharedAccount>(Signer::address_of(&resource_signer))) {
+            move_to(
+                &resource_signer,
+                SharedAccount {
+                    share_record: Vector::empty<State>(),
+                    signer_capability: resource_signer_cap,
+                }
+            )
+        };
+
+        let shared_account = borrow_global_mut<SharedAccount>(Signer::address_of(&resource_signer));
+
+        let i = 0;
+        let cumulative_numerator = 0;
+        let cumulative_denominator = 1;
+
+        while (i < Vector::length(&addresses)) {
+            let num = *Vector::borrow(&numerators, i);
+            let denom = *Vector::borrow(&denominators, i);
+            let addr = *Vector::borrow(&addresses, i);
+            assert!(Account::exists_at(addr), Errors::invalid_argument(EACCOUNT_NOT_FOUND));
+            assert!(num >= 0 && denom > 0, Errors::invalid_argument(EINVALID_INPUT));
+
+            let common_denom = find_least_common_denominator(cumulative_denominator, denom);
+            cumulative_numerator = cumulative_numerator * (common_denom / cumulative_denominator) + num * (common_denom / denom);
+            cumulative_denominator = common_denom;
+
+            Vector::push_back(&mut shared_account.share_record, State { share_holder: addr, numerator: num, denominator: denom });
+            i = i + 1;
+        };
+
+        assert!(cumulative_numerator == cumulative_denominator, Errors::invalid_argument(EINVALID_NUMERATOR_DENOMINATOR_COMBINATIONS));
+        Signer::address_of(&resource_signer)
+    }
+
+    // Disperse all available balance to addresses in the shared account
+    public(script) fun disperse<CoinType>(resource_addr: address) acquires SharedAccount {
+        assert!(exists<SharedAccount>(resource_addr), Errors::invalid_argument(EINVALID_SIGNER));
+
+        let total_balance = Coin::balance<CoinType>(resource_addr);
+        assert!(total_balance > 0, Errors::limit_exceeded(EINSUFFICIENT_BALANCE));
+
+        let shared_account = borrow_global<SharedAccount>(resource_addr);
+        let resource_signer = Account::create_signer_with_capability(&shared_account.signer_capability);
+
+        let i = 0;
+        while (i < Vector::length(&shared_account.share_record)) {
+            let state = Vector::borrow(&shared_account.share_record, i);
+            let current_amount = state.numerator * total_balance / state.denominator;
+            let current_balance = Coin::balance<CoinType>(resource_addr);
```

Comment:
we can use Coin::balance outside of the loop and update the current_balance = current_balance - current_amount in each iteration.